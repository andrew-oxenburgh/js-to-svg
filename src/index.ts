import SVG = require('svgson')
export type Origin = { x: number, y: number }
export type Size = { width: number, height: number }

function useSomeShortCuts(attr: object, propName: string, tx: string[][]) {
   if (typeof attr[propName] == 'object') {
      for (const off in tx) {
         const transform = tx[off]
         attr[transform[0]] = attr[propName][transform[1]]
      }
      delete attr[propName]
   }
}

function createSvgObject(attr: object, children: object[]): object {
   const DEF = {
      xmlns: 'http://www.w3.org/2000/svg',
   }
   return {
      'name': 'svg',
      'type': 'element',
      'value': '',
      'attributes': {
         ...DEF,
         ...attr
      },
      'children': children
   }
}

const g = (children: object[], attr = {}): object => {
   return elem('g', attr, children)
}
const elem = (name: string, attr = {}, children: object[] = []): object => {
   return {
      'name': name,
      'type': 'element',
      'value': '',
      'attributes': attr,
      'children': children
   }
}

const path = (attr: object, children: object[] = []): object => {
   const DEF = {
   }
   const attributes = {...DEF, ...attr}

   return elem('path', attributes, children)
}
/**
 * Rectangle
 * in @attr will accept a size, an origin, or a rectangle
 */
const rect = (attr: object, children: object[] = []): object => {
   const DEF = {
   }
   useSomeShortCuts(attr, 'size', [
      ['width', 'width'],
      ['height', 'height'],
   ])

   useSomeShortCuts(attr, 'origin', [
      ['x', 'x'],
      ['y', 'y'],
   ])

   useSomeShortCuts(attr, 'rect', [
      ['x', 'x'],
      ['y', 'y'],
      ['width', 'width'],
      ['height', 'height'],
   ])

   const attributes = {...DEF, ...attr}

   return elem('rect', attributes, children)
}
const circle = (attr: object, children: object[] = []): object => {
   const DEF = {
      fill: 'none',
   }

   useSomeShortCuts(attr, 'center', [
      ['cx', 'x'],
      ['cy', 'y'],
   ])

   if(typeof attr['square'] === 'object'){
      const square = attr['square']
      attr['cx'] = square.x + (square?.side / 2)
      attr['cy'] = square.y + (square?.side / 2)
      attr['r'] = square?.side / 2
      delete attr['square']
   }

   const attributes = {...DEF, ...attr}
   return elem('circle', attributes, children)
}
const line = (attr: object, children: object[] = []): object => {
   const DEF = {
      'stroke-linecap': 'round'
   }
   useSomeShortCuts(attr, 'from', [
      ['x1', 'x'],
      ['y1', 'y'],
   ])
   useSomeShortCuts(attr, 'to', [
      ['x2', 'x'],
      ['y2', 'y'],
   ])

   const attributes = {...DEF, ...attr}
   return elem('line', attributes, children)
}

const text = (attr = {}, text: string): object => {
   return {
      'name': 'text',
      'type': 'element',
      'value': '',
      'attributes': attr,
      'children': [
         {
            'name': '',
            'type': 'text',
            'value': text,
            'attributes': {},
            'children': []
         }
      ]
   }
}
const animate = (attr = {}): object => {
   return elem(
      'animate',
      attr
   )
}
module.exports = {
   createSvgObject,
   g,
   elem,
   path,
   rect,
   circle,
   line,
   text,
   animate,
   stringify: SVG.stringify
}
