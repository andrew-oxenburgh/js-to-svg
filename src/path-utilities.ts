import {Arc, Point} from './index'

export function moveA(point: Point | string): string {
   if (typeof point === 'string') {
      return point
   }
   return `M${point.x},${point.y}`
}

export function moveR(point: Point | string): string {
   if (typeof point === 'string') {
      return point
   }
   return `m${point.x},${point.y}`
}

export function complete(): string {
   return 'Z'
}

export function arcA(arc: Arc | string): string {
   if (typeof arc === 'string') {
      return arc
   }
   return `A${arc.rx},${arc.ry},${arc.angle},${arc.largeArcFlag},${arc.sweepFlag},${arc.x},${arc.y}`
}

/**
 * Accept an even numbered length of an array of points, or a string which passes straight through it
 * @param points
 */
export function quadraticA(points: Point[] | string): string {
   if (typeof points === 'string') {
      return points
   }
   if (points.length < 1) {
      return ''
   }
   if (points.length % 2) {
      throw 'Quadratic Error: requires an even number of points'
   }
   const res = 'Q' + points.reduce((acc: string, point: Point) => {
      acc += `${point.x},${point.y} `
      return acc
   }, '')
   const removeTrailingComma = res.slice(0, res.length - 1)
   return removeTrailingComma
}
