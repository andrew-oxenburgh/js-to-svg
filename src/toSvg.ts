let SVG = require('svgson');

type Origin = { x: number, y: number };

type Size = { width: number, height: number };
const line = (attr: object): object => {
   const DEF = {
      'stroke-linecap': 'round'
   }
   const attributes = {...DEF, ...attr}
   return elem('line', attributes, [])
}

const elem = (name: string, attr = {}, children: object[] = [], value = ''): object => {
   return {
      "name": name,
      "type": "element",
      "value": "",
      "attributes": attr,
      "children": children
   }
}

const text = (attr = {}, text: string): object => {
   return {
      "name": "text",
      "type": "element",
      "value": "",
      "attributes": attr,
      "children": [
         {
            "name": "",
            "type": "text",
            "value": text,
            "attributes": {},
            "children": []
         }
      ]
   }
}

const circle = (attr: {}): object => {
   const DEF = {
      fill: 'none',
   };

   const attributes = {...DEF, ...attr}
   return elem('circle', attributes, [])
}

const rect = (attr: {}): object => {
   const DEF = {
      fill: 'none',
   };

   const attributes = {...DEF, ...attr}
   return elem('rect', attributes, [])
}

const g = (children: object[], attr = {}): object => {
   return elem('g', attr, children)
}

function createSvgObject(children: object[], origin: Origin, size: Size): object {
   return {
      "name": "svg",
      "type": "element",
      "value": "",
      "attributes": {
         width: size.width + '',
         height: size.height + '',
         viewBox: `${origin.x} ${origin.y} ${size.height} ${size.width}`,
         xmlns: "http://www.w3.org/2000/svg"
      },
      "children": children
   };
}

module.exports = {
   createSvgObject,
   g,
   circle,
   line,
   elem,
   text,
   rect,
   stringify: SVG.stringify
}
