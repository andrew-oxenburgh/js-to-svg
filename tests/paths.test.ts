'use strict'
import {expectSnapshot} from "./test-utils";
import {Point, Arc, Line} from "../src";

const toSvg = require('../src/index')
describe('fixing-path', () => {
   test('heart', () => {
      let centerLine: Line = {x: 500};
      let bulgeLine: Line = {y: 300}
      let leftLine: Line = {x: 100}
      let rightLine: Line = {x: 900}
      let topLine: Line = {y: 900}
      let startingPoint: Point = {x: 100, ...bulgeLine};

      let dipPoint: Point = {...centerLine, ...bulgeLine};
      let left: Line = {x: 900};

      let leftUpper: Arc = {rx: 200, ry: 200, angle: 0, largeArcFlag: 0, sweepFlag: 1, ...dipPoint};
      let rightUpper: Arc = {...leftUpper, ...left};
      let rightLower: Point[] = [{...rightLine, y: 600}, {...centerLine, ...topLine}];
      let leftLower: Point[] = [{...leftLine, y: 600}, startingPoint];

      let moveTo = toSvg.moveA(startingPoint);
      let arc1 = toSvg.arcA(leftUpper);
      let arc2 = toSvg.arcA(rightUpper)
      let quad1 = toSvg.quadraticA(rightLower);
      let quad2 = toSvg.quadraticA(leftLower);
      let z = toSvg.complete();

      let path = [
         moveTo,
         arc1,
         arc2,
         quad1,
         quad2,
         z
      ];
      const json = [
         toSvg.path({
            id: 'heart',
            d: path.join(''),
            fill: 'red'
         }),
      ]
      expectSnapshot(json, 'complex path')
   })
})
