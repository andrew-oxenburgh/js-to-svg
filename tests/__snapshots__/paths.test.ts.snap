// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixing-path heart 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>complex path</title>
  <path id="heart" d="M100,300A200,200,0,0,1,500,300A200,200,0,0,1,900,300Q900,600,500,900Q100,600,100,300Z" fill="red"/>
</svg>"
`;
